### PR TITLE
Implementing Library Shelves as a UI Upgrade

### PR DESCRIPTION
**Pull request by felixOlesen**
**Description**
Implementing the library shelves requires:
- gibbonLibraryShelf table and gibbonLibraryShelfItem relational tables to be implemented. _Done ✅_
- add, addProcess, edit, editProcess, delete, and deleteProcess basic forms and process files to be made for CRUD of the libraryShelves. _Done ✅_
- Setting up necessary ajax files with respect to different form inputs requiring them. _Done ✅_
- Writing LibraryShelfGateway and LibraryShelfItemGateway files for querying the database. _Done ✅_
- Writing up a twig template for rendering the library shelves in the Library > Browse Library page. _Done ✅_

**Motivation and Context**
The Library > Browse Library page was implemented with a simple table, displaying all the information of library items present in gibbonLibraryItem. We were approached by a librarian to try and make the catalog browsing more exciting for other users. Therefore, we were asked to implement a netflix-like rails design in the form of library shelves.

**How Has This Been Tested?**
Tested Locally

**Screenshots**
Manage Library Shelves
Screenshot 2024-02-29 at 12 00 07

Add Library Shelves
Screenshot 2024-02-29 at 12 00 40

Edit Library Shelves
Screenshot 2024-02-29 at 12 00 55

Browse Library Shelves
Screenshot 2024-02-29 at 12 21 58

Advanced Search of Library Items
Screenshot 2024-02-29 at 12 18 31

### CODE REVIEW

File: CHANGEDB.php [hunk 1]
```diff
@@ -793,6 +793,10 @@
...
+INSERT INTO `gibbonPermission` (`gibbonRoleID` ,`gibbonActionID`) VALUES ('001', (SELECT gibbonActionID FROM gibbonAction JOIN gibbonModule ON (gibbonAction.gibbonModuleID=gibbonModule.gibbonModuleID) WHERE gibbonModule.name='Library' AND gibbonAction.name='Manage Library Shelves'));end
+CREATE TABLE `gibbonLibraryShelfItem` ( `gibbonLibraryShelfItemID` INT(6) UNSIGNED ZEROFILL NOT NULL AUTO_INCREMENT , `gibbonLibraryItemID` INT(10) UNSIGNED ZEROFILL NOT NULL, `gibbonLibraryShelfID` INT(6) UNSIGNED ZEROFILL NOT NULL, PRIMARY KEY (`gibbonLibraryShelfItemID`)) ENGINE = InnoDB CHARSET=utf8;end
 INSERT INTO `gibbonNotificationEvent` (`event`, `moduleName`, `actionName`, `type`, `scopes`, `active`) VALUES ('First Aid Qualification Expiry', 'Staff', 'Manage Staff', 'Core', 'All', 'Y');end
```
Comment (SKuipers) [inline CHANGEDB.php:800]: Re-ordering the CHANGEDB, as all new entries need added at the bottom. I'll accept & commit the suggested change, no action needed on your part.

```suggestion
INSERT INTO `gibbonNotificationEvent` (`event`, `moduleName`, `actionName`, `type`, `scopes`, `active`) VALUES ('First Aid Qualification Expiry', 'Staff', 'Manage Staff', 'Core', 'All', 'Y');end
CREATE TABLE `gibbonLibraryShelf` ( `gibbonLibraryShelfID` INT(6) UNSIGNED ZEROFILL NOT NULL AUTO_INCREMENT , `name` VARCHAR(90) NOT NULL, `active` ENUM('Y','N') NOT NULL DEFAULT 'N', `field` VARCHAR(90) NOT NULL, `fieldValue` VARCHAR(90), `type` ENUM('MANUAL','AUTOMATIC') NOT NULL DEFAULT 'MANUAL', `sequenceNumber` INT(6) NOT NULL, PRIMARY KEY (`gibbonLibraryShelfID`)) ENGINE = InnoDB CHARSET=utf8;end
INSERT INTO `gibbonAction` (`gibbonModuleID`, `name`, `precedence`, `category`, `description`, `helpURL`, `URLList`, `entryURL`, `entrySidebar`, `menuShow`, `defaultPermissionAdmin`, `defaultPermissionTeacher`, `defaultPermissionStudent`, `defaultPermissionParent`, `defaultPermissionSupport`, `categoryPermissionStaff`, `categoryPermissionStudent`, `categoryPermissionParent`, `categoryPermissionOther`) VALUES((SELECT gibbonModuleID FROM gibbonModule WHERE name='Library'), 'Manage Library Shelves', 0, 'Catalog', 'Allows user to manage shelves displaying library items for library browsing.', NULL, 'library_manage_shelves.php, library_manage_shelves_add.php, library_manage_shelves_edit.php, library_manage_shelves_delete.php', 'library_manage_shelves.php', 'Y', 'Y', 'Y', 'N', 'N', 'N', 'N', 'Y', 'N', 'N', 'N');end
INSERT INTO `gibbonPermission` (`gibbonRoleID` ,`gibbonActionID`) VALUES ('001', (SELECT gibbonActionID FROM gibbonAction JOIN gibbonModule ON (gibbonAction.gibbonModuleID=gibbonModule.gibbonModuleID) WHERE gibbonModule.name='Library' AND gibbonAction.name='Manage Library Shelves'));end
CREATE TABLE `gibbonLibraryShelfItem` ( `gibbonLibraryShelfItemID` INT(6) UNSIGNED ZEROFILL NOT NULL AUTO_INCREMENT , `gibbonLibraryItemID` INT(10) UNSIGNED ZEROFILL NOT NULL, `gibbonLibraryShelfID` INT(6) UNSIGNED ZEROFILL NOT NULL, PRIMARY KEY (`gibbonLibraryShelfItemID`)) ENGINE = InnoDB CHARSET=utf8;end
```

File: CHANGEDB.php [hunk 2]
```diff
@@ -794,5 +794,9 @@
 UPDATE `gibbonAction` SET category='Learn' WHERE gibbonModuleID=(SELECT gibbonModuleID FROM gibbonModule WHERE name='School Admin') AND name='Individual Needs Settings';end
 ALTER TABLE `gibbonPerson` CHANGE `departureReason` `departureReason` VARCHAR(100) NOT NULL;end
 INSERT INTO `gibbonNotificationEvent` (`event`, `moduleName`, `actionName`, `type`, `scopes`, `active`) VALUES ('First Aid Qualification Expiry', 'Staff', 'Manage Staff', 'Core', 'All', 'Y');end
+CREATE TABLE `gibbonLibraryShelf` ( `gibbonLibraryShelfID` INT(6) UNSIGNED ZEROFILL NOT NULL AUTO_INCREMENT , `name` VARCHAR(90) NOT NULL, `active` ENUM('Y','N') NOT NULL DEFAULT 'N', `field` VARCHAR(90) NOT NULL, `fieldValue` VARCHAR(90), `type` ENUM('MANUAL','AUTOMATIC') NOT NULL DEFAULT 'MANUAL', `sequenceNumber` INT(6) NOT NULL, PRIMARY KEY (`gibbonLibraryShelfID`)) ENGINE = InnoDB CHARSET=utf8;end
```
Comment (SKuipers) [inline CHANGEDB.php:797]:
```suggestion
CREATE TABLE `gibbonLibraryShelf` ( `gibbonLibraryShelfID` INT(6) UNSIGNED ZEROFILL NOT NULL AUTO_INCREMENT , `name` VARCHAR(90) NOT NULL, `active` ENUM('Y','N') NOT NULL DEFAULT 'N', `field` VARCHAR(90) NOT NULL, `fieldValue` VARCHAR(90), `type` ENUM('Manual','Automatic') NOT NULL DEFAULT 'Manual', `sequenceNumber` INT(6) NOT NULL, PRIMARY KEY (`gibbonLibraryShelfID`)) ENGINE = InnoDB CHARSET=utf8;end
```

File: modules/Library/library_manage_shelves_add.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            $row->addYesNo('active')
+                ->required();
+
+        $row = $form->addRow();
+            $row->addLabel('type', __('Fill Option'));
+            $row->addSelect('type')
+                ->required()
+                ->fromArray([
+                    'automatic' => __('Automatic'),
+                    'manual' => __('Manual')
```
Comment (SKuipers) [inline modules/Library/library_manage_shelves_add.php:71]:
```suggestion
                    'Automatic' => __('Automatic'),
                    'Manual' => __('Manual')
```

File: modules/Library/library_manage_shelves_add.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                ])
+                ->placeholder('Please select...')
+                ->selected($urlParams['type']);
+
+        $form->toggleVisibilityByClass('automatic')->onSelect('type')->when('automatic');
```
Comment (SKuipers) [inline modules/Library/library_manage_shelves_add.php:76]:
```suggestion
        $form->toggleVisibilityByClass('automatic')->onSelect('type')->when('Automatic');
```

File: modules/Library/library_manage_shelves_addProcess.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    $shelfItems = isset($_POST['addItems'])? explode(',', $_POST['addItems']) : [];
+    //$gibbonLibraryTypeID = isset($_POST['gibbonLibraryTypeID'])? $_POST['gibbonLibraryTypeID'] : '';
+    $gibbonLibraryTypeID = $_POST['gibbonLibraryTypeID'];
+    
+    if(!empty($shelfItems) && $data['type'] == 'manual') {
+        foreach($shelfItems as $item) {
+            $itemGateway->insertShelfItem($item, $gibbonLibraryShelfID);
+        }
+    } else if(!empty($gibbonLibraryTypeID) && $data['type'] == 'automatic') {
```
Comment (SKuipers) [inline modules/Library/library_manage_shelves_addProcess.php:77]:
```suggestion
    } else if(!empty($gibbonLibraryTypeID) && $data['type'] == 'Automatic') {
```

File: modules/Library/library_manage_shelves_edit.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    $row = $form->addRow();
+        $row->addLabel('type', __('Fill Option'));
+        $row->addTextField('type')->setValue($values['type'])->readOnly();
+
+    $row = $form->addRow();
+        $row->addLabel('field', __('Category'));
+        $row->addTextField('field')->setValue($values['field'])->readOnly();
+
+    if($values['type'] == 'AUTOMATIC') {
```
Comment (SKuipers) [inline modules/Library/library_manage_shelves_edit.php:82]:
```suggestion
    if($values['type'] == 'Automatic') {
```

File: modules/Library/library_manage_shelves_edit.php [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        $row = $form->addRow();
+        $row->addLabel('fieldValue', __('Sub-Category'));
+        $row->addTextField('fieldValue')->setValue($values['fieldValue'])->readOnly()
+            ->required();
+
+    } elseif($values['type'] == 'MANUAL') {
```
Comment (SKuipers) [inline modules/Library/library_manage_shelves_edit.php:88]:
```suggestion
    } elseif($values['type'] == 'Manual') {
```

File: modules/Library/library_manage_shelves_add.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        
+        $row = $form->addRow()->addClass('autoFill');
+            $row->addLabel('fieldValue', __('Sub-Category'));
+            $row->addSelect('fieldValue')
+                ->fromArray($categories['subCategory'])
+                ->chainedTo('field', $categories['subCategoryChained'])
+                ->placeholder('Please select...')
+                ->selected($urlParams['fieldValue']);
+
+        $form->toggleVisibilityByClass('manual')->onSelect('type')->when('manual');
```
Comment (SKuipers) [inline modules/Library/library_manage_shelves_add.php:105]:
```suggestion
        $form->toggleVisibilityByClass('manual')->onSelect('type')->when('Manual');
```

File: modules/Library/library_manage_shelves_addProcess.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    // Create the substitute
+    $gibbonLibraryShelfID = $shelfGateway->insert($data);
+
+    $shelfItems = isset($_POST['addItems'])? explode(',', $_POST['addItems']) : [];
+    //$gibbonLibraryTypeID = isset($_POST['gibbonLibraryTypeID'])? $_POST['gibbonLibraryTypeID'] : '';
+    $gibbonLibraryTypeID = $_POST['gibbonLibraryTypeID'];
+    
+    if(!empty($shelfItems) && $data['type'] == 'manual') {
```
Comment (SKuipers) [inline modules/Library/library_manage_shelves_addProcess.php:73]:
```suggestion
    if(!empty($shelfItems) && $data['type'] == 'Manual') {
```

**Review by SKuipers — CHANGES_REQUESTED** [review]
Hi Felix, this is looking amazing and well thought out, I'm really looking forward to seeing it in production. The only suggested changes I've noted are in the case consistency of your `type` field. In some places, it's used all lowercase, and other places it's used all uppercase. For consistency, both with itself and with the codebase as a whole, I've suggested changing it to Title Case in all locations. Aside from that, very keen to merge and test this out 😃

**Comment by felixOlesen** [comment]
Hi Sandra, thanks very much for the feedback. The suggested changes look good to me so I've committed them in. 👍 I'm glad that it's all looking ok and excited to know it'll be in production soon! 😁

**Review by SKuipers — APPROVED** [review]
Thanks for the quick response and changes. I'm going to merge this in now and test it out 🎉

**Comment by SKuipers** [comment]
Hi @felixOlesen Thanks again for your amazing work on this new feature, it has now gone live in ICHK, just in time for Book Week 🎉 

Post-merge, I made some tweaks and changes before pushing to production, found [in this commit](https://github.com/GibbonEdu/core/commit/c1ea66766076ee85aba808acd0c0fe290bb50f49).

- Removed the border radius, since books are square, it felt a little visually strange to see them with rounded corners.
- Lowered the transition duration, so the hover-over feels snappier.
- Changed the library search from grid to flex with flex-wrap, since the grid was causing books to overlap each other.
- Fixed a PHP notice when there are no shelves.
- Added some upper page size limits, to reduce page load time, since there is no pagination.
- Added some default library shelves for Top 20 and New Titles, so even if there are no shelves, the browse page isn't blank. These could ideally be worked into a type of automatic shelf, so they could be reordered and added/removed as desired.
- I found a tweak that can be made to the Google Books API url, to retrieve a higher quality image for some books (if ome is available), so that the book thumbnails aren't as blurry.

An additional consideration for development:

- Item status should include all copies of an item. Otherwise it may appear that the book is unavailable when it really is available in other copies.

Thanks again! It is lovely to see these changes in production just in time for book week.

Screenshot 2024-03-04 at 08-46-12 ICHK - Gibbon - Library

**Comment by felixOlesen** [comment]
Hi Sandra, these tweaks sound awesome! Thanks very much for sorting it out. I've started development on those auto-shelves, I'll make a PR draft once I have something more substantial. I'm so glad it's up and running for book week! 😁

**Pull request merged**